File: soln-tree-problems/binary-tree-maximum-path-sum.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/binary-tree-maximum-path-sum/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max_sum=Integer.MIN_VALUE;
    
    public int maxPathSum(TreeNode root) {
        if(root==null)
            return 0;
        dfs(root);
        return max_sum;
    }
        
    public int dfs(TreeNode root){
        if(root==null)
            return 0;
        
        int left=dfs(root.left);
        int right=dfs(root.right);
        
        int straightPath=Math.max(Math.max(left,right)+root.val,root.val);
        int rootPath=Math.max(left+right+root.val,root.val);
        max_sum=Math.max(max_sum,Math.max(straightPath,rootPath));
        
        return straightPath;
    }
}

File: Python/soln-binary-tree-problems/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    max_sum=-1000
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0;
        
        def dfs(root):
            if root is None:
                return 0
            
            left=dfs(root.left)
            right=dfs(root.right)
            
            straightPath=max(max(left,right)+root.val,root.val)
            summ=left+root.val+right
            rootPath=max(summ,root.val)
            self.max_sum=max(self.max_sum,straightPath,rootPath)
            
            return straightPath
            
        dfs(root)
        return self.max_sum
